manuals/lnkinfo.1
.Dd June 25, 2026
.Dt LNKINFO 1
.Os
.Sh NAME
.Nm lnkinfo
.Nd determines information about a Windows Shortcut File (LNK)
.Sh SYNOPSIS
.Nm lnkinfo
.Op Fl c Ar codepage
.Op Fl hvV
.Ar source
.Sh DESCRIPTION
.Nm lnkinfo
is a utility to determine information about a Windows Shortcut File (LNK)
.Pp
.Nm lnkinfo
is part of the
.Nm liblnk
package.
.Nm liblnk
is a library to access the Windows Shortcut File (LNK) format
.Pp
.Ar source
is the source file.
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl c Ar codepage
codepage of ASCII strings, options: ascii, windows-874, windows-932, \
windows-936, windows-949, windows-950, windows-1250, windows-1251, windows-1252 \
(default), windows-1253, windows-1254, windows-1255, windows-1256, windows-1257 \
or windows-1258
.It Fl h
shows this help
.It Fl v
verbose output to stderr
.It Fl V
print version
.El
.Sh ENVIRONMENT
None
.Sh FILES
None
.Sh EXAMPLES
.Bd -literal
# lnkinfo Calculator.lnk
lnkinfo 20140831
.sp
Windows Shortcut information:
	Contains a link target identifier
	Contains a description string
	Contains a working directory string
	Contains an environment variables block
.sp
Link information:
	Creation time			: Aug 10, 2004 16:54:24.000000000 UTC
	Modification time		: Aug 04, 2004 14:00:00.000000000 UTC
	Access time			: Jun 26, 2006 10:36:41.703125000 UTC
	File size			: 114688 bytes
	File attribute flags		: 0x00000020
		Should be archived (FILE_ATTRIBUTE_ARCHIVE)
	Drive type			: Fixed (3)
	Drive serial number		: 0xc868f004
	Volume label			: System
	Local path			: C:\\WINDOWS\\system32\\calc.exe
	Description			: @%SystemRoot%\\system32\\shell32.dll,-22531
	Working directory		: C:\\WINDOWS\\system32
	Environment variables location	: %SystemRoot%\\system32\\calc.exe
.sp
Link target identifier:
	Shell item list
		Number of items		: 5
.sp
	Shell item: 1
		Item type		: Root folder
		Class type indicator	: 0x1f (Root folder)
		Shell folder identifier	: 20d04fe0-3aea-1069-a2d8-08002b30309d
		Shell folder name	: My Computer
.sp
	Shell item: 2
		Item type		: Volume
		Class type indicator	: 0x2f (Volume)
		Volume name		: C:\\
.sp
	Shell item: 3
		Item type		: File entry
		Class type indicator	: 0x31 (File entry: Directory)
		Name			: WINDOWS
		Modification time	: Dec 17, 2006 20:55:44
		File attribute flags	: 0x00000010
			Is directory (FILE_ATTRIBUTE_DIRECTORY)
	Extension block: 1
		Signature		: 0xbeef0004 (File entry extension)
		Long name		: WINDOWS
		Creation time		: Feb 01, 2006 21:49:46
		Access time		: Dec 17, 2006 20:55:48
.sp
	Shell item: 4
		Item type		: File entry
		Class type indicator	: 0x31 (File entry: Directory)
		Name			: system32
		Modification time	: Nov 26, 2006 19:56:54
		File attribute flags	: 0x00000030
			Is directory (FILE_ATTRIBUTE_DIRECTORY)
			Should be archived (FILE_ATTRIBUTE_ARCHIVE)
	Extension block: 1
		Signature		: 0xbeef0004 (File entry extension)
		Long name		: system32
		Creation time		: Feb 01, 2006 21:49:46
		Access time		: Dec 17, 2006 20:49:46
.sp
	Shell item: 5
		Item type		: File entry
		Class type indicator	: 0x32 (File entry: File)
		Name			: calc.exe
		Modification time	: Aug 04, 2004 14:00:00
		File attribute flags	: 0x00000020
			Should be archived (FILE_ATTRIBUTE_ARCHIVE)
	Extension block: 1
		Signature		: 0xbeef0004 (File entry extension)
		Long name		: calc.exe
		Creation time		: Aug 10, 2004 16:54:24
		Access time		: Jun 26, 2006 10:36:42
.sp
Distributed link tracking data:
	Machine identifier		: hostname
	Droid volume identifier		: f6084ef7-7b6a-41e7-a286-e376c0ffd656
	Droid file identifier		: b7d0e438-4422-48b0-be9d-b4bc8794cef6
	Birth droid volume identifier	: f6084ef7-7b6a-41e7-a286-e376c0ffd656
	Birth droid file identifier	: b7d0e438-4422-48b0-be9d-b4bc8794cef6
.sp
.Ed
.Sh DIAGNOSTICS
Errors, verbose and debug output are printed to stderr when verbose output \
\-v is enabled.
Verbose and debug output are only printed when enabled at compilation.
.Sh AUTHORS
.An Joachim Metz <joachim.metz@gmail.com>
.Sh BUGS
Please report bugs of any kind on the project issue tracker: \
https://github.com/libyal/liblnk/issues
.Sh COPYRIGHT
Copyright (C) 2009-2026, Joachim Metz <joachim.metz@gmail.com>.
.sp
This is free software; see the source for copying conditions.
There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A \
PARTICULAR PURPOSE.


ChangeLog
TODO
* add thread safety, continue with data block

* lnkinfo: : should be before \t
* check behavior of liblnk_file_link_refers_to_file() seems to block certain time values from being printed.
* what about trailing data in xp lnk files ?
* what about shell items
* pylnk add functions
* rename liblnk_file_copy_link_target_identifier_data to liblnk_file_get_link_target_identifier_data ?

File format:
* Ignore the location information (ForceNoLinkInfo)
* what is controlled by the has location info?

Checks
* compare environment variables string and unicode string
* compare icon location data string, string and unicode string
  - data string not necessary equal to string in data block, due to expansion of %VAR%

Refactor
* volume information read into separate subfunction
* network share information read into separate subfunction

Debug
* print descriptive name of data_block_signature
* print values of additional extra blocks
  - console properties data block
  - console codepage data block
  - shim layer data block
  - metadata property store data block
  - shell item identifiers list data block
* remaining known folder identifiers in test set

Library
* create API functions to access link information values

API
* add functions
  - for icon location block ?
  - for special and/or known folder identifier
* remove liblnk_file_get_link_target_identifier_data

20090905
* see `git log' for more recent change log
* initial version



lnktools/lnkinfo.c
/*
 * Shows information obtained from a Windows Shortcut File (LNK).
 *
 * Copyright (C) 2009-2026, Joachim Metz <joachim.metz@gmail.com>
 *
 * Refer to AUTHORS for acknowledgements.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

#include <common.h>
#include <file_stream.h>
#include <system_string.h>
#include <types.h>

#if defined( HAVE_FCNTL_H ) || defined( WINAPI )
#include <fcntl.h>
#endif

#if defined( HAVE_IO_H ) || defined( WINAPI )
#include <io.h>
#endif

#if defined( HAVE_STDLIB_H ) || defined( WINAPI )
#include <stdlib.h>
#endif

#if defined( HAVE_UNISTD_H )
#include <unistd.h>
#endif

#include "info_handle.h"
#include "lnktools_getopt.h"
#include "lnktools_libcerror.h"
#include "lnktools_libclocale.h"
#include "lnktools_libcnotify.h"
#include "lnktools_liblnk.h"
#include "lnktools_output.h"
#include "lnktools_signal.h"
#include "lnktools_unused.h"

info_handle_t *lnkinfo_info_handle = NULL;
int lnkinfo_abort                  = 0;

/* Signal handler for lnkinfo
 */
void lnkinfo_signal_handler(
      lnktools_signal_t signal LNKTOOLS_ATTRIBUTE_UNUSED )
{
	libcerror_error_t *error = NULL;
	static char *function    = "lnkinfo_signal_handler";

	LNKTOOLS_UNREFERENCED_PARAMETER( signal )

	lnkinfo_abort = 1;

	if( lnkinfo_info_handle != NULL )
	{
		if( info_handle_signal_abort(
		     lnkinfo_info_handle,
		     &error ) != 1 )
		{
			libcnotify_printf(
			 "%s: unable to signal info handle to abort.\n",
			 function );

			libcnotify_print_error_backtrace(
			 error );
			libcerror_error_free(
			 &error );
		}
	}
	/* Force stdin to close otherwise any function reading it will remain blocked
	 */
#if defined( WINAPI ) && !defined( __CYGWIN__ )
	if( _close(
	     0 ) != 0 )
#else
	if( close(
	     0 ) != 0 )
#endif
	{
		libcnotify_printf(
		 "%s: unable to close stdin.\n",
		 function );
	}
}

/* The main program
 */
#if defined( HAVE_WIDE_SYSTEM_CHARACTER )
int wmain( int argc, wchar_t * const argv[] )
#else
int main( int argc, char * const argv[] )
#endif
{
	const char *description = \
		"Use lnkinfo to determine information about a Windows Shortcut File (LNK).";

	lnktools_option_t options[ ] = {
		{ 'c', "codepage", "codepage of ASCII strings, options: ascii, windows-874, windows-932, windows-936, windows-949, windows-950, windows-1250, windows-1251, windows-1252 (default), windows-1253, windows-1254, windows-1255, windows-1256, windows-1257 or windows-1258" },
		{ 'h', NULL, "shows this help" },
		{ 'v', NULL, "verbose output to stderr" },
		{ 'V', NULL, "print version" },
		{ 0, "source", "the source file" },
	};
	system_character_t options_string[ 32 ];

	int result                          = 0;
	liblnk_error_t *error               = NULL;
	system_character_t *option_codepage = NULL;
	system_character_t *source          = NULL;
	char *program                       = "lnkinfo";
	system_integer_t option             = 0;
	int number_of_options               = (int) ( sizeof( options ) / sizeof( lnktools_option_t ) );
	int verbose                         = 0;

#if defined( __MINGW32__ ) && defined( HAVE_MINGW_BINMODE )
	_setmode( _fileno( stdout ), _O_BINARY );
	_setmode( _fileno( stderr ), _O_BINARY );
#endif

	libcnotify_stream_set(
	 stderr,
	 NULL );
	libcnotify_verbose_set(
	 1 );

	if( libclocale_initialize(
	     "lnktools",
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize locale values.\n" );

		goto on_error;
	}
	if( lnktools_output_initialize(
	     _IONBF,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize output settings.\n" );

		goto on_error;
	}
	lnktools_output_version_fprint(
	 stdout,
	 program );

	if( lnktools_getopt_get_options_string(
	     options,
	     number_of_options,
	     options_string,
	     32 ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to determine options string.\n" );

		goto on_error;
	}
	while( ( option = lnktools_getopt(
	                   argc,
	                   argv,
	                   options_string ) ) != (system_integer_t) -1 )
	{
		switch( option )
		{
			case (system_integer_t) '?':
			default:
				fprintf(
				 stderr,
				 "Invalid argument: %" PRIs_SYSTEM "\n",
				 argv[ optind - 1 ] );

				lnktools_getopt_usage_fprint(
				 stdout,
				 program,
				 description,
				 options,
				 number_of_options );

				return( EXIT_FAILURE );

			case (system_integer_t) 'c':
				option_codepage = optarg;

				break;

			case (system_integer_t) 'h':
				lnktools_getopt_usage_fprint(
				 stdout,
				 program,
				 description,
				 options,
				 number_of_options );

				return( EXIT_SUCCESS );

			case (system_integer_t) 'v':
				verbose = 1;

				break;

			case (system_integer_t) 'V':
				lnktools_output_copyright_fprint(
				 stdout );

				return( EXIT_SUCCESS );
		}
	}
	if( optind == argc )
	{
		fprintf(
		 stderr,
		 "Missing source file.\n" );

		lnktools_getopt_usage_fprint(
		 stdout,
		 program,
		 description,
		 options,
		 number_of_options );

		return( EXIT_FAILURE );
	}
	source = argv[ optind ];

	libcnotify_verbose_set(
	 verbose );
	liblnk_notify_set_stream(
	 stderr,
	 NULL );
	liblnk_notify_set_verbose(
	 verbose );

	if( info_handle_initialize(
	     &lnkinfo_info_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize info handle.\n" );

		goto on_error;
	}
	if( option_codepage != NULL )
	{
		result = info_handle_set_ascii_codepage(
		          lnkinfo_info_handle,
		          option_codepage,
		          &error );

		if( result == -1 )
		{
			fprintf(
			 stderr,
			 "Unable to set ASCII codepage in info handle.\n" );

			goto on_error;
		}
		else if( result == 0 )
		{
			fprintf(
			 stderr,
			 "Unsupported ASCII codepage defaulting to: windows-1252.\n" );
		}
	}
	if( info_handle_open_input(
	     lnkinfo_info_handle,
	     source,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to open source file.\n" );

		goto on_error;
	}
	if( info_handle_file_fprint(
	     lnkinfo_info_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to print file information.\n" );

		goto on_error;
	}
	if( info_handle_close_input(
	     lnkinfo_info_handle,
	     &error ) != 0 )
	{
		fprintf(
		 stderr,
		 "Unable to close info handle.\n" );

		goto on_error;
	}
	if( info_handle_free(
	     &lnkinfo_info_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to free info handle.\n" );

		goto on_error;
	}
	return( EXIT_SUCCESS );

on_error:
	if( error != NULL )
	{
		libcnotify_print_error_backtrace(
		 error );
		libcerror_error_free(
		 &error );
	}
	if( lnkinfo_info_handle != NULL )
	{
		info_handle_free(
		 &lnkinfo_info_handle,
		 NULL );
	}
	return( EXIT_FAILURE );
}

